### PR TITLE
Refactoring FFEvaluate

### LINKED ISSUE #391: FFEvaluate energy of torsion term does not match with OpenMM
Get aniline (Ph-NH2) from https://raw.githubusercontent.com/Acellera/parameterise-testset/master/reference/m051/input.mol2

Calculate energy of aniline with HTMD:
```python
from htmd.parameterization.ffmolecule import FFMolecule
from htmd.parameterization.fftype import FFTypeMethod
from htmd.parameterization.ffevaluate import FFEvaluate

mol = FFMolecule('input.mol2', method=FFTypeMethod.GAFF2)
res = FFEvaluate(mol).evaluate(mol.coords[:,:,0])
print(res)

typemap = mol._prm.writeFrcmod(mol._rtf, 'mol.frcmod')
mol.write('mol.mol2', typemap=typemap)
``` 
```
{'elec': -2.051702647788332, 'vdw': 3.9122669768070497, 'bond': 0.77021301246485774, 'angle': 0.4833635420588632, 'dihedral': 1.612904528740123, 'improper': 0.0, 'total': 4.7270454122825614}
```

Generate *prmtop* and *inpcrd* files
```bash
$ cat tleap.inp
loadAmberParams mol.frcmod
MOL = loadMol2 mol.mol2
saveAmberParm MOL mol.prmtop mol.inpcrd
quit
$ tleap -f tleap.inp
```

Calculate energy of aniline with *ParmEd/OpenMM*:
```python
import parmed
from simtk import openmm 
from simtk import unit

prmtop = parmed.amber.LoadParm('mol.prmtop', xyz='mol.inpcrd')
system = prmtop.createSystem()
integrator = openmm.LangevinIntegrator(300 * unit.kelvin, 1 / unit.picoseconds, 2 * unit.femtoseconds)
platform = openmm.Platform.getPlatformByName('CPU')
context = openmm.Context(system, integrator, platform)
context.setPositions(prmtop.positions)
res = parmed.openmm.energy_decomposition(prmtop, context)
print(res)
```
```
{'bond': 0.7702162476656408, 'angle': 0.48337638371315556, 'dihedral': 2.8833159558052577, 'nonbonded': 1.8605609115871495, 'total': 5.997469498771204}
```

The energy of dihedral term does not match:
```
HTMD: 1.612904528740123
OpenMM: 2.8833159558052577
```

### LINKED ISSUE #393: Dihedral angle calculation is inconsistent
At https://github.com/Acellera/htmd/blob/master/htmd/parameterization/ffevaluate.py
```python
    def evaluateTorsion(coords, phis):
        pos1 = coords[0, :]
        pos2 = coords[1, :]
        pos3 = coords[2, :]
        pos4 = coords[3, :]

        r12 = pos1 - pos2
        r23 = pos2 - pos3
        r34 = pos3 - pos4

        r12 = r12.flatten()
        r23 = r23.flatten()
        r34 = r34.flatten()

        A = np.cross(r12, r23)
        B = np.cross(r23, r34)
        C = np.cross(r23, A)

        rA = 1. / np.linalg.norm(A)
        rB = 1. / np.linalg.norm(B)
        rC = 1. / np.linalg.norm(C)

        B = B.dot(rB)

        cos_phi = A.dot(B) * rA
        sin_phi = C.dot(B) * rC

        phi = - math.atan2(sin_phi, cos_phi)
...
```

At https://github.com/Acellera/htmd/blob/master/htmd/parameterization/phi.py
```python
def getPhi(coords, atoms):
    coords = np.atleast_3d(coords)
    pos1 = coords[atoms[0], :, 0]
    pos2 = coords[atoms[1], :, 0]
    pos3 = coords[atoms[2], :, 0]
    pos4 = coords[atoms[3], :, 0]

    r12 = pos1 - pos2
    r23 = pos2 - pos3
    r34 = pos3 - pos4

    A = np.cross(r12, r23)
    B = np.cross(r23, r34)
    C = np.cross(r23, A)

    rA = 1. / np.linalg.norm(A)
    rB = 1. / np.linalg.norm(B)
    rC = 1. / np.linalg.norm(C)

    B = B * rB

    cos_phi = (np.dot(A, B)) * rA
    sin_phi = (np.dot(C, B)) * rC

    phi = - math.atan2(sin_phi, cos_phi)
...
```

Identical code, except:
```python
B = B.dot(rB)
```
versus
```python
B = B * rB
```

### PR DESCRIPTION
**Pull request by raimis**
- [x] Fix (partially) #393 
- [x] Test energies versus OpenMM
  -  [x] CHARMM
     - [x] Depends on  #398 
  -  [x] AMBER
     - [x] Depends on #392
     - [x] Fix #391
- [x] Clean code
- [x] Documentation

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -52,6 +52,10 @@ install:
   - conda install --file package/htmd-deps/DEPENDENCIES -y
   - python package/htmd-deps/write_meta_yaml.py package/htmd-deps
 
+  # TODO: this is test-time dependencies for FFEvaluate. Move into the right place
+  - conda install openmm
+  - conda install -c omnia parmed
```
Comment (j3mdamas) [inline .travis.yml]: We need to decide if we want HTMD to depende on these packages or not

Comment (raimis) [inline .travis.yml]: This is just a temporary solution for testing.

**Comment by j3mdamas** [comment]
@raimis do you want me to review this?

**Comment by raimis** [comment]
@j3mdamas not yet! First, I want to finish #423, so nothing gets broken.

**Comment by raimis** [comment]
I let you know, when it is ready!

**Comment by raimis** [comment]
Now this depends on #423

**Comment by raimis** [comment]
@j3mdamas it is time for this one!

**Comment by raimis** [comment]
OK, @j3mdamas ! All the stuff squashed into one commit.

**Comment by j3mdamas** [comment]
Thanks so much @raimis!

**Comment by raimis** [comment]
@j3mdamas 

There is a problem *parameterize* tests: they don't catch errors!

**Comment by j3mdamas** [comment]
@raimis You have to give me more information

**Comment by j3mdamas** [comment]
Thanks for reporting @raimis! This should fix is: https://github.com/Acellera/htmd/commit/b2fc68f94c8114fa42e6ad170b390b152e20addc
I had to quarantine 3 tests. They are very long tests, not run on Travis anyway. One of them I haven't tested, because it's too long, and the other two fail on dihedral energy checking. It seems like it fluctuates between two values, nothing could be done there, and since the tests are temporary, I won't lose more time with it. I reversed back the energies.txt changes and all the work is done by `assertAlmostEqual`, which I put to 4 decimal places. For the parameter files, I think it's fine. I think that the Impropers in the energies report may not be good for this, but we can always fix it with a formatted print. Let me know if this is good.

**Comment by j3mdamas** [comment]
wow, apparently the `MASS` records of atom types can also swap order in `mol.rtf` files: https://travis-ci.org/Acellera/htmd/jobs/269369228

Fixing that by treating it file your hack for .frcmod files

**Comment by j3mdamas** [comment]
OK, the tests should fail now, as you expect.

**Comment by j3mdamas** [comment]
One of the tests (`h2o2_dihed_fix_restart`) fails on py36 (but not py35), on the dihedral energy. I'm going to skip it on that build.

**Comment by j3mdamas** [comment]
OK, passed.

**Comment by j3mdamas** [comment]
OK; just reviewed. Small changes, and as soon as passes Travis, I will merge into master. @cuzzo87 @raimis

**Comment by j3mdamas** [comment]
@cuzzo87 these were merged, I think you can run what you wanted to

**Comment by raimis** [comment]
Thanks @j3mdamas for polishing the tests almost to perfection... :D

**Comment by j3mdamas** [comment]
Irony? :P

**Pull request merged**